File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
     bool DFS_Cycle(vector<vector<int>>& Adj, vector<int>& status, int i, vector<int>& Inchain)
    {
        status[i]=1;
        Inchain[i]=1;
        for(int j: Adj[i]){
            if(status[j]==0){
                bool B = DFS_Cycle(Adj, status, j, Inchain);

                if(B==true){
                    return true;
                }
                
            }
            else{ 
                    if(Inchain[j]==1){
                         return true;
                    }
            }
            
        }
        Inchain[i]=0;
        return false;

    }
bool Cycle_Detection(vector<vector<int>> Adj,int V){
        vector<int>status(V,0);
        for(int i=0;i<V;i++){
            if(status[i]==0){
                vector<int>Inchain(V,0);
                bool B = DFS_Cycle(Adj,status,i,Inchain);
                if(B==true){
                    
                    return true;
                }
                
            }
            
            
        }
        return false;
}

    
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>>Adj(numCourses);
        int a ,b;
        for(int i=0;i<prerequisites.size();i++){
             
             a = prerequisites[i][0];
             b = prerequisites[i][1];

            Adj[a].push_back(b);

        
      }
      vector<int>ans(numCourses,0);

      vector<int>InDeg(numCourses,0);

      for(int i=0;i<numCourses;i++){
        
        for(int j=0;j<Adj[i].size();j++){
            InDeg[Adj[i][j]]++;
        }
        
    }


    vector<int>vec;


      if(!Cycle_Detection(Adj,numCourses)){
            
          queue<int>q;
        for(int i=0;i<numCourses;i++){
            if(InDeg[i]==0){
                q.push(i);
                vec.push_back(i);
            }
        }
        while(!q.empty()){
            int x = q.front();
            q.pop();
            
            for(int y: Adj[x]){
                InDeg[y]--;
                if(InDeg[y]==0){
                    q.push(y);
                    vec.push_back(y);
                }
            }
        }
        }

        reverse(vec.begin(),vec.end());

        return vec;
        
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int V = numCourses;
        
        
       ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
        
        for(int i=0;i<prerequisites.length;i++){
            int u = prerequisites[i][0];
            int v = prerequisites[i][1];
            
            adj.get(v).add(u);
            
            
        }
        
        int []indegree = new int[V];
        
        for(int i=0;i<V;i++){
            for(int y:adj.get(i)){
                indegree[y]++;
            }
        }
        
        Queue<Integer> q = new LinkedList<>();
        
        for(int i=0;i<V;i++){
            if(indegree[i]==0){
                q.add(i);
            }
        }
        
        
        int[] ans = new int[V];
        int i = 0;
        
        while(!q.isEmpty()){
            int x = q.poll();
            ans[i++]=x;
            
            for(int y:adj.get(x)){
                indegree[y]--;
                if(indegree[y]==0){
                    q.add(y);
                }
            }
            
            
        }
        if (i != V) return new int[0];
        return ans;
    }
}